### DISCUSSION TITLE
How do pocketbase handles multiple transactions?

### DISCUSSION THREAD

**Discussion by mrbu-solutions** [pocketbase/pocketbase#2816] [Q&A] [answered]
Sorry I'm new to backend programming.
 
Should the backend logic handles the queuing when doing CRUD in pocketbase, or pocketbase will handle it for the backend?

The documentation shows that it calls directly pocketbase operations after receiving the request. 

So if I have 1 million users using my app that calls my API, and my API executes pocketbase' create action, how could pocketbase handle the sequence of this transaction?

**Comment by ganigeorgiev** [comment, marked as answer]
SQLite in WAL allows only a single writer at a time, meaning that multiple write requests will queue up.

> So if I have 1 million users using my app that calls my API, and my API executes pocketbase' create action, how could pocketbase handle the sequence of this transaction?

I'm not sure that I understand the question and what do you mean by _"sequence of this transaction"_.  If you are asking for the order of the operations, there is no guarantee that the requests will be processed in the same order as they were sent because each request is handled concurrently in its own goroutine and when (and where) each goroutine will execute is decided by the go runtime/scheduler.

**Reply by mrbu-solutions** [reply]
Would your answer mean that I don't need to have my own scheduler/queuer to handle concurrent request because goroutine will handle it? If that so, then this is great. Thank you very much for your time in answering my question.

**Reply by ganigeorgiev** [reply]
Yes, users usually shouldn't need to to anything in addition to handle concurrent requests.

**Reply by dschissler** [reply]
I once found that due to complicated logic that a PB model event was creating a transaction where a (relatively) long shell / remote execution was happening.  This quickly failed once I went from my dev machine to the early test server.  It was obvious once I thought of it, but the flow of my application made it hard to think about.  It happened like this:

I have a collection in my database that everything revolves around.  It became too complex to put logic in the field update and insert hooks (due to transaction and remote processes), and so I moved it into a special form for that operation.  It's possible that I could get it to work in the model hook again, but it's much easier to only have that operation happening via a custom route that calls the form.  So that means that I cannot allow direct access to the collection via PB insert and update API.

I think that each collection needs to be looked at separately to see if it's really worth the hassle of managing complex processes via pocketbase model events.  It's always there if you need it.  So I'm using a hybrid approach where everything uses PB API with filters unless that is actually going to cost me time and frustration.

I could refactor and refactor, but in my case, why?  For this important collection I'm accessing it via custom SQL view anyways.  This also allows me to hide other collections without figuring out crazy tangled slow API rule creep.  I'll add that I'm still using `expand` for supplemental data, and it works good.

I'm still figuring out PB design patterns, but I think that if I ever find myself needing a transaction for a collection then I'll 1) put that logic in a form (modeled off of the pretty decent PB internal forms), 2) Create a custom route that binds to that form.  Here I can allow extra options for the form that are not bound using any struct tag.  This adds flexibility.  3) turn off the PB API for update and insert of that collection, 4) Create a custom SQL view for that collection with joins, etc.
